### DISCUSSION TITLE
New platform development. Help needed!

### DISCUSSION THREAD

**Discussion by hanke-cnc** [Paciente8159/uCNC#772] [Q&A]
I have seen your strong programming skills and talent. I need your email to communicate with you. I believe you shouldn't waste your time and talent on adapting to various MCUs; you need to focus on one thing.

**Comment by Paciente8159** [comment]
You can join the discord channel and send me a private message there. I'm usually online.

**Reply by hanke-cnc** [reply]
Okay, let's continue the discussion here. I'm from China, and first I need to state that I'm not a company, but an individual, a CNC enthusiast. I'm just looking for a partner and hope you can maintain a separate version of uCNC for me. I want a pure CNC motion controller, with a fixed chip, without unnecessary features (I consider things like web pages, Bluetooth, SD card, etc., to be boring and a waste of time).

**Reply by hanke-cnc** [reply]
My main job is as a CNC operator. Three years ago, in my spare time, I used GRBL to DIY a desktop CNC. I am very interested in CNC, or more precisely, in numerical control systems. I taught myself some electronics and started making GRBL circuit boards for sale. A year ago, I developed an offline CNC controller that is low-cost, priced at only $71, and modern, featuring a 7-inch touchscreen with a GPU that supports 3D tool paths. It has been selling well to many DIY CNC enthusiasts. Since I am not a dedicated programmer, I am concerned about its future development, so I need a partner to handle the lower-level machine (motion controller).

**Reply by Paciente8159** [reply]
Do you have a specific MCU/chip you are considering for this job?

**Reply by hanke-cnc** [reply]
> Do you have a specific MCU/chip you are considering for this job?

RP2350 or STM32H750VBT6, they are powerful enough to be prepared for the next 10 years, and they are inexpensive, priced at only $1.7 in China. This allows circuit boards to be sold at a very low cost. I am very interested in your multi-core work, with one core handling communication and another handling motion control. It's fantastic! Very pure!

**Reply by Paciente8159** [reply]
The RP2350 is already as an [initial support PR](https://github.com/Paciente8159/uCNC/pull/770).
I should also start development for the stm32h7 family until the end of the year (I hope)

**Reply by hanke-cnc** [reply]
> The RP2350 is already as an [initial support PR](https://github.com/Paciente8159/uCNC/pull/770). I should also start development for the stm32h7 family until the end of the year (I hope)

It seems that you are not willing to collaborate with me on a closed-source CNC control project or to maintain an RP2350 project for me on a paid basis. It's not an official commercial project, and there are no constraints. I'm just using my spare time to make circuit boards and sell them.

**Reply by hanke-cnc** [reply]
A few days ago, I purchased an RP2040 development board. It seems that the long path issue you mentioned caused me to waste a lot of time without successfully compiling. After checking your documentation, I enabled the long path for compiling and flashing the program, but it didn't run on the actual machine. I sent several "?" commands to it, and it replied with a very large and scary number in either the machine coordinates or WCO. As for the RP2350, I couldn't find it in the VSCode compile options. My programming skills are quite weak.

**Reply by Paciente8159** [reply]
Define very long and scary number in coordinates or wco.
Seems like a normal status message

**Reply by Paciente8159** [reply]
Also note that the rp2350 is still somewhat in it's early stages. Things like the emulated eeprom do not work

**Reply by hanke-cnc** [reply]
Okay, I will first purchase an RP2350 development board and wait. I have a plan for a 4.3-inch touchscreen designed for CNC beginners. It runs Linux+Qt and is very affordable, only $14.

**Reply by hanke-cnc** [reply]
Hi, I would like to inquire if it is too difficult to have some industrial controller functions, such as G84 tapping, threading on a lathe, automatic tool change, radius compensation, etc. This is why I hope to pay to find a partner, because I am not very good at programming and cannot take it further.

**Reply by Paciente8159** [reply]
> Hi, I would like to inquire if it is too difficult to have some industrial controller functions, such as G84 tapping, threading on a lathe, automatic tool change, radius compensation, etc. This is why I hope to pay to find a partner, because I am not very good at programming and cannot take it further.

I'll try to answer the best way I can.
- G84 is not currently implemented. It's possible to be implement taking what was done with G33 as reference. It's not implemented yet because I don't have a machine to test it out. I need to build a small lathe type jig to conduct proper tests on it.
- Threading on a lathe is implemented using uCNC extension module G33. This modules HAS VERY LITTLE testing and is just a partial implementation since G33.1 is not implemented. Again the above mentioned lathe jig and some proper testing would be fundamental.
- uCNC is prepared for automatic tool change but this feature depends a great deal on how the tool changer works and is implemented. All tools have a startup and shutdown function that can be modified to run code (for example the code to do the tool switch). This gave me an idea. I can implement a generic case were a user can define a gcode file to run on startup and another on shutdown. If the tool changer can be fully controlled with Gcode then this should do the trick.
- Radius compensation is probably the only one that I hardly think it can implemented. The problem with G84 is that it must be implemented in a per tool path. Each tool path can contain anything from a few lines of Gcode to thousands of lines. On small device like these MCU there is just not enough RAM to perform that. The only way this might be achieved would be to have that in a file, the parser go though the whole file and basically generate a second file with the compensated path. For that you might as well do it in a PC alot more easier and faster. Also this is more a task for the CAM itself then the machine (I think).

**Reply by hanke-cnc** [reply]
So cool, it's already running. RP2350 and UCNC are still a bit unfamiliar to me, but I'm getting familiar with them.
![AA](https://github.com/user-attachments/assets/4d7862f3-c09f-440b-beca-5dfc3b5dcea4)

**Reply by Paciente8159** [reply]
Beware that the RP2350 core is still at an early stage of development. I had very little changes to test it. The Arduino framework itself seems a bit incomplete too. Things like EEPROM emulation, Multicore and SPI with DMA support are not working.

**Reply by hanke-cnc** [reply]
OK, I understand. I have a question: Can it use USB for virtual serial communication? This would make it convenient for users to upgrade the firmware. Or can the RP2350 use serial port 0 for firmware upgrades?

**Comment by Paciente8159** [comment]
Serial communications run both on UART and USB.
As far as I know you need to use USB to upload firmware.

**Reply by hanke-cnc** [reply]
> Serial communications run both on UART and USB. As far as I know you need to use USB to upload firmware.

Hello, I am planning to make a circuit board for him based on the RP2350 and test it. There might be some issues later on. Currently, the issue is that the $0 parameter does not allow me to set it to 3. I consulted ChatGPT and modified the function in `grbl_settings.c`. Now, it can be set to 3. However, when I run the command `G1 G91 X9999 F12000` to measure the X-axis, it only outputs at 30kHz, which is unacceptable for both servo users and modern users. On grblHAL, the STM32F407 can reach 400kHz, and the H7 can reach 700kHz. I believe the M33 core is at least more powerful than the M4 core. You might have overlooked something.

```
$0=3.000
$2=0
$3=0
$4=0
$5=0
$6=0
$7=0
$10=1
$11=0.200
$12=0.002
$13=0
$20=0
$21=0
$22=0
$23=0
$24=0
$25=0
$26=250
$27=2.000
$30=1000
$31=0
$32=0
$81=0.000
$100=800.000
$101=200.000
$102=200.000
$110=25500.000
$111=500.000
$112=500.000
$120=10.000
$121=10.000
$122=10.000
$130=200.000
$131=200.000
$132=200.000
```

**Reply by Paciente8159** [reply]
That settings is limited by the the macro F_STEP_MAX that has different default limits for all architectures.
If not defined it will default to 64KHz in the case of the RP2350
To change that you have to add that setting.
For example you can add this inside your boardmap_rpi_pico2.h file

`#define F_STEP_MAX 200000UL // This sets to 200KHz. you can specify other values`

**Reply by Paciente8159** [reply]
Also note that the RP2350 documentation seems to be quite incomplete.
This is taken from the SDK datasheet:
![{521956F7-BBEE-4801-BC6B-DA0110E22781}](https://github.com/user-attachments/assets/97017789-7bea-4ed8-9794-a72b013054a1)
But if you go to the RP2350 datasheet you only find 5 points and I did not found any info on the tick block.

So in practical terms you are (for now) stuck to a resolution of 1us for the timer. That will mean that the realistically speaking you should not expect good resolution step increments above 100KHz stepping rate.

**Reply by hanke-cnc** [reply]
Even after adding `#define F_STEP_MAX 200000UL` in `boardmap_rpi_pico2.h`, I still only obtained pulses of around 35 kHz. Even with 100 kHz, I can't accept it. Although 100-200 kHz is completely sufficient for CNC, I can't sell it because users will inquire about this parameter. They believe the higher the better, and you can't change their mindset. I hope for around 400-500 kHz. I'm not very skilled in programming. Are you saying that the rp2350 cannot output 400-500 kHz? Should I expect and wait for the stm32H7's ucnc? It's really a pity. The rp2350 is multi-core with an M33 core and is cheap enough, which is very tempting for me.

**Reply by Paciente8159** [reply]
Did you undo the changes you did to the `grbl_settings.c` file? If not restore the original file. Modifying code there might lead to unwanted behavior.

I'll only be able to test the limits of the RP2350 tomorrow. As far as I can get from the documentation, yes I am for now locked to a timer resolution no better then 1µs. It might be possible to reach 500Khz steprate but will be a mess because of the resolution.
With that limitation (while accelerating or decelerating) you will see a step jump from 250KHz to 500KHz and that most probably will cause huge vibrations and motor stalling.

Maybe when further clarifications on how change the timers to higher resolution are available it might be possible.

**Reply by hanke-cnc** [reply]
I only modified the following in `grbl_settings.c`:

```c
// id 0 conversion from us to frequency
if (id == 0)
{
    if (value < 3)
    {
        return STATUS_MAX_STEP_RATE_EXCEEDED;
    }
}
```

Then, I added `#define F_STEP_MAX 200000UL` in `boardmap_rpi_pico2.h`, but it seems to have no effect and remains at 35kHz. I have not checked the RP2350's documentation and resources; it seems unsuitable for making into a product. I consulted the `stm32H750VBT6` manual, and it has one high-resolution timer (maximum resolution 2.1 ns), two 32-bit timers, and many other timers. I can purchase the `stm32H750VBT6` for $1.7. I might wait for your `stm32h7`'s uCNC.

**Reply by Paciente8159** [reply]
Yes I have plans to integrate the STM32H7 soon.
I'm waiting for the RP2350 to mature a bit more and started working on integrating Arduino UNO R4 with the Renesas RA4M1 (but that is going to take a bit more time since probably).
I'll probably going to get a BigTechTree Octopus Pro board to kickstart STM32H7 before the end of the year.

**Reply by Paciente8159** [reply]
Ok so I tested the master code on the RP2350.
I changed the F_STEP_MAX [here](https://github.com/Paciente8159/uCNC/blob/61e50b69385e3051673594d273c7f19667675643/uCNC/src/hal/mcus/rp2350/mcumap_rp2350.h#L51) and kind of worked.

The max step rate would somehow fluctuated around125KHz (I guess mostly from rounding to the closest value possible because of the timer resolution of Mhz).
Even if the timer resolution issue can be overcome I think (stable) stepping rates above 200KHz for a single axis on this MCU will probably require some in depth optimizations.

**Reply by hanke-cnc** [reply]
好的，
It seems RP2350 is not suitable for me,
I will wait for your BigTechTree Octopus Pro,
I have checked the parameters of BigTechTree Octopus Pro,
it is H723's,
Does H723 have a hardware trigonometric function accelerator? Looks cool?
In addition,
I used PayPal to donate a cup of coffee to you,
I am Chinese,
first time using PayPal to pay,
it shows payment was successful on my side,
I think you should have received it.

**Reply by Paciente8159** [reply]
> I am Chinese,
> first time using PayPal to pay,

Busted my ass laughing :laughing:
I did receive it thanks.

I've been looking at boards for a couple of months now and trying to figure which is best for future development

Here is my candidate list:
BIGTREETECH BTT Octopus Pro v1.1
BIGTREETECH BTT SKR 3
BIGTREETECH BTT SKR 3 EZ
BIGTREETECH Stealthy Hi-Speed Solution
Makerbase MKS SKIPR (nevermind this uses a F407)

**Reply by Paciente8159** [reply]
Out of curiosity I tested the step speed of a single axis on my blackpill (snt32f401) running at 84MHz and got a very stable 266KHz step rate...not bad. Don't know if I can squeeze a bit more out of that or not.

I will try the mks nano 3.1 (running at 168MHz) latter this evening to see what I will get out of it.

**Reply by hanke-cnc** [reply]
I am currently using/selling the grblhal STM32F407 control board. The grblhal for the F401 operates at 300 kHz, while the F407/F446 operates at 400 kHz. I am not someone who blindly pursues this parameter; 400 kHz is sufficient for promotional purposes.

For me, the STM32H750VBT6 is the best choice, costing only $1.7. Secondly, the H723 is $2.85.

**Reply by Paciente8159** [reply]
Just to let you know I have received the board a did a quick test with the current code I have. On a single axis I achieved about 650KHz.
But this was a quick and dirty test and the code is still in an early stage and is still very raw.
I have a lot of work right now. I will only resume development, probably in the early 2025.

**Reply by hanke-cnc** [reply]
> Just to let you know I have received the board a did a quick test with the current code I have. On a single axis I achieved about 650KHz. But this was a quick and dirty test and the code is still in an early stage and is still very raw. I have a lot of work right now. I will only resume development, probably in the early 2025.

Sorry, I was refreshing the page to see if you had the latest reply. Since your reply wasn't at the bottom, I missed your message.

650kHz is great, as long as it's above 400kHz, it's okay. It's just a selling point for a product. I've interacted with many DIY CNC buyers, so I know what they need.

I will wait for you and look forward to working with you.

Recently, something happened that made me more determined to seek collaboration with professionals. It started when I modified the original GRBLHAL code, changing the function of the <Start> button to send a message to my host computer for processing when triggered. Since I'm not a professional programmer, I wrote incorrect code that made it easy to trigger <Start>. Initially, only two users reported the issue to me, but they couldn't accurately describe the problem, and I couldn't reproduce it. Not everyone experienced this issue until recently, when a user triggered <Start> by inserting a USB drive into the controller, which caught my attention. This incident was really terrible, and it made me realize that professional tasks should be handled by professionals.

**Reply by Paciente8159** [reply]
Making code is not the hardest part. Testing and debuging for all possible cenarios way harder and time consuming. Mistakes is part of the process.

Did you reached trerjeio from GrblHAL for support?
What made you switch from using grblHal?

**Reply by hanke-cnc** [reply]
I haven't given up on using grblHAL. I'm currently using it, and it's great, but I'm unable to add or modify the features that users need. I don't have the capability, and reading grblHAL even makes me feel dizzy.

Trerjeio is very nice. Whenever I find and report a bug about grblHAL, it gets resolved quickly. However, I can't submit my personal needs to him because he's very busy, and I feel that his development philosophy for grblHAL is different from mine. I don't understand why he spends so much time maintaining various MCUs and features like SD, Ethernet, and web pages, which I don't need. Is it because it's a non-profit project?

For me, choosing a cost-effective and powerful chip like the H750, based on the needs of DIY CNC enthusiasts, would be ideal. It can be used for at least 10 years without needing replacement, allowing the author to focus more time and energy on maintaining and developing it.

**Reply by Paciente8159** [reply]
Your development philosophy is very different from Terjeio or mine for that matter.
Your goal stands on making the most profitable return from the machines or systems you want to build.

I cannot speak for Terjeio but I assume that he maintains several MCU and features for the same reason I do. His goal is not to develop a firmware/features that only fits one specific board, product or application but that can benefit a more broad spectrum of users, needs and hardware, instead of  a particular use case or a very peculiar user need. In my case I started this project and maintain it just because I like it.

Most users of these types of projects are hobby machinists and home CNC users, or with limited budgets/hardware resources, not industry machines. 

I believe Terjeio is very busy yes, but so am I. I've willingly accepted working on the core for the STM32H7 because it was a MCU family that was on my roadmap for quite some time. But if a feature does not make sense to me I will not waist my time on it.

**Reply by hanke-cnc** [reply]
I am very sad that you seem to think I am a profit-driven businessman. 😖

**Reply by hanke-cnc** [reply]
Terjeio is also not interested in my ideas. I was originally a DIY CNC enthusiast, and I still am. I'm not a company; I sell it in my spare time. I just want to make it better. I'm not a programmer and can't add or modify code in the existing grblhal code, so I'm looking for a partner.

**Reply by hanke-cnc** [reply]
My goal is not industrial machines either, as I don't have the capability or funds to compete with industrial controllers. My target is also DIY CNC enthusiasts. Most DIY CNC players don't write code, so selling them stable, affordable, and reliable control boards is also a form of help. I believe that something profitable is a good thing; otherwise, how would the author have the energy and time to maintain it? Therefore, I also think Terjeio is selfless and great.

**Reply by Paciente8159** [reply]
> otherwise, how would the author have the energy and time to maintain it?

How about just because the author enjoys it? Somethings are just priceless. The freedom to do a thing just because you like doing it without strings attached, deadlines or obligations has no measurable value. As soon as you remove that freedom from the process it becomes like any other regular job.

**Reply by hanke-cnc** [reply]
Alright, I completely respect and understand what you mean.

**Comment by Paciente8159** [comment]
Ok. While I make up my mind about the board I'll will order a generic STM32H750 development board to kickstart development. It should take about 2 to 3 weeks to arrive.

**Reply by hanke-cnc** [reply]
> Ok. While I make up my mind about the board I'll will order a generic STM32H750 development board to kickstart development. It should take about 2 to 3 weeks to arrive.

I still want to try to get along with you for a while, starting with the STM32H7. I will test and provide feedback on any bugs. Please notify me if your STM32H7 uCNC is ready.

**Reply by hanke-cnc** [reply]
I tested your UCNC's S-curve acceleration about 1 or 2 years ago by placing a cup of water on my machine and using G1 motion. It seemed like it didn't work, so I gave up on it.

**Reply by Paciente8159** [reply]
> I tested your UCNC's S-curve acceleration about 1 or 2 years ago by placing a cup of water on my machine and using G1 motion. It seemed like it didn't work, so I gave up on it.

The pseudo S curves were modified about a year ago (Jul 2023). Previous to that the s curve was applied in a different way and had some issues. So I don't know if you used the most recent version or not.

**Reply by Paciente8159** [reply]
> > Ok. While I make up my mind about the board I'll will order a generic STM32H750 development board to kickstart development. It should take about 2 to 3 weeks to arrive.
> 
> I still want to try to get along with you for a while, starting with the STM32H7. I will test and provide feedback on any bugs. Please notify me if your STM32H7 uCNC is ready.

Thanks. I'll let you know. All possible tests carried out on any platform are more then welcome. That is a very time consuming task.

**Reply by hanke-cnc** [reply]
Hi, is there any update on the H7-uCNC?

**Reply by Paciente8159** [reply]
I have a very basic working version working but pretty much not tested with several things missing (like SPI, EEPROM emulation).
For now it seems to be working with the GPIO generation, timers USB and UART. I have no support for low power drivers also.

**Reply by hanke-cnc** [reply]
I don't need SPI.

I don't want to use flash EEPROM emulation to save settings, as it can lead to errors, which I cannot tolerate. I previously used EEPROM emulation on grblHAL, but if written in a short time, grbl would return strange errors. So I used the EEPROM plugin, specifically the 24LC16 chip, to save settings: https://github.com/grblHAL/Plugin_EEPROM/

This is the error I once reported to terjeio: https://github.com/grblHAL/core/issues/336

**Reply by Paciente8159** [reply]
External EEPROM like that also need I2C (also not implemented and tested).
And other EEPROM's or flash memories use SPI.
SPI also might be used for screens and other stuff.

I will gradually implement and test each feature to some extent at least, before releasing this core. I might not implement EEPROM emulation (I can drop that one for now, but not the others).

**Reply by hanke-cnc** [reply]
I have another question: Can your uCNC open a second serial port to receive commands? I used to use Terjeio's keypad plugin to open a serial port that only receives G-code to implement the handwheel function for GRBL.

This is a low-cost 3D printed enclosure handwheel I made and sold to DIY CNC enthusiasts. It costs only $14 and is very cool.
![AA](https://github.com/user-attachments/assets/abc6e102-687a-476e-bd94-2f051f812b4e)

**Reply by Paciente8159** [reply]
Yes you can have multiple ports sending commands to the controller.

**Reply by Paciente8159** [reply]
I should run some tests this weekend. Still several things need to be developed and test, but ~~USB and~~ UART should be working.
TODO:

- USB
- SPI (with and without DMA)
- I2C
- EEPROM emulation (if possible)

**Reply by hanke-cnc** [reply]
Understood.

**Reply by hanke-cnc** [reply]
How is the progress going? I noticed you made some changes. There is code for H7 in grblhal, which you might find helpful: https://github.com/dresco/STM32H7xx. Sorry, I'm not good with code, so I can't help much, but I did sponsor you a cup of coffee.

**Reply by Paciente8159** [reply]
I did get usb to work but not the way I wanted. I would like to go mostly baremetal, but I might simply use arduino infraestructure and then tey to slowly replace it wit baremetal code. Lets see...

**Reply by hanke-cnc** [reply]
I have had a question for a long time and have no one to ask. If we use USB virtual serial ports for communication, and USB has CRC checks, does this mean that no additional checks are needed? I use serial communication without any checks and seemingly haven't encountered any communication errors (the control board and the host computer's motherboard are very close, connected using a 100mm cable). I searched online and found some answers:

[It’s two different concepts. The USB's CRC ensures that the data transmitted by it is correct, but you cannot guarantee that the data provided to it is correct for you.]

**Reply by Paciente8159** [reply]
USB has internal protocol error checking. Since Grbl protocol is a text protocol you don't need CRC. A corrupted text will lead to an invalid command (usually). So yes It might be the case that the error makes a X transform to an Y.

**Reply by Paciente8159** [reply]
By the way USB is now also working.

**Reply by hanke-cnc** [reply]
This is not a problem, just a discussion. I have reviewed the documentation of a Chinese industrial controller, as well as the very expensive Masso controller (docs.masso.com.au). Should we make all inputs configurable features?

The configurable input documentation for Masso is here: https://docs.masso.com.au/wiring-and-setup/setup-and-calibration/list-of-configurable-inputs

**Reply by hanke-cnc** [reply]
I'm not sure if it will make things more complicated. I think if the user configures a certain input as the X-axis limit, then let ucnc handle it itself. If not, then send the information to the host computer for processing. In short, things are divided into ucnc handling itself and sending to the host computer for processing. What do you think of my idea?

**Reply by Paciente8159** [reply]
I'm not sure if I understood your question.
µCNC works over the Grbl message protocol. When a user defines a X limit pin the board might or not make use of that information (controlled by settings $21 and $22). Regardless of the board actually doing something with that input (to perform homing or limit/crash detection) the status of all special inputs (like control pins, limits and probe) is always available in the status message. What the program does or not with that information depends only on the software side of the machine. There are some caveats.

Other generic inputs can be configured to detect stuff, synchronize gcode excution, custom actions, etc...depending on a module implementation for that.

The delay between the actual limit trigger and the actual software handle is undetermined. Depends on how frequently the status is being pulled, transmission speeds, and several other factors.

**Reply by hanke-cnc** [reply]
I'm just reading its documentation and have some questions. It seems that all inputs can be defined/configured in the host software, such as tool setting, ESTOP emergency stop, and limit switches, and can be set as normally open or normally closed. Additionally, I found that each axis surprisingly requires three signals (for example, Z-axis positive limit, Z-axis negative limit, and Z-axis mechanical origin), which is quite confusing.

**Reply by hanke-cnc** [reply]
I'm not sure why it was designed this way. Why does each axis require three limit signals? Perhaps the original GRBL intended to design it this way but was limited by the Arduino's pins.

**Reply by hanke-cnc** [reply]
I remember operating a machining center with a Fanuc system, which can indeed trigger hard limits in both the positive and negative directions. However, what I don't understand is why it would trigger hard limits when soft limits are already in place, especially since it uses absolute value servo motors. It's perplexing.

**Reply by Paciente8159** [reply]
Failsafe?! Software limits depend on user input and hardware limits protect against human error.

**Reply by hanke-cnc** [reply]
> Failsafe?! Software limits depend on user input and hardware limits protect against human error.

Hello, I have already downloaded the stm32H750 firmware you compiled, and it works fine. However, when I tried to compile it myself, I encountered errors. Did you forget to upload some files?

|-- Wire @ 1.0.0
Building in release mode
Compiling .pio\build\STM32H7-Generic-STM32H750\SrcWrapper\src\HAL\stm32yyxx_hal.c.o
Compiling .pio\build\STM32H7-Generic-STM32H750\SrcWrapper\src\HAL\stm32yyxx_hal_adc.c.o
Compiling .pio\build\STM32H7-Generic-STM32H750\SrcWrapper\src\HAL\stm32yyxx_hal_adc_ex.c.o
<command-line>: fatal error: variant_WeActMiniH7xx.h: No such file or directory

*******************************************************************************
* Looking for variant_WeActMiniH7xx.h dependency? Check our library registry!
*
* CLI  > platformio lib search "header:variant_WeActMiniH7xx.h"
* Web  > https://registry.platformio.org/search?q=header:variant_WeActMiniH7xx.h
*
*******************************************************************************

Compiling .pio\build\STM32H7-Generic-STM32H750\SrcWrapper\src\HAL\stm32yyxx_hal_can.c.o
compilation terminated.
Compiling .pio\build\STM32H7-Generic-STM32H750\SrcWrapper\src\HAL\stm32yyxx_hal_cec.c.o
<command-line>: fatal error: variant_WeActMiniH7xx.h: No such file or directory

*******************************************************************************
* Looking for variant_WeActMiniH7xx.h dependency? Check our library registry!
*
* CLI  > platformio lib search "header:variant_WeActMiniH7xx.h"
* Web  > https://registry.platformio.org/search?q=header:variant_WeActMiniH7xx.h
*
*******************************************************************************

compilation terminated.
Compiling .pio\build\STM32H7-Generic-STM32H750\SrcWrapper\src\HAL\stm32yyxx_hal_comp.c.o
*** [.pio\build\STM32H7-Generic-STM32H750\SrcWrapper\src\HAL\stm32yyxx_hal.c.o] Error 1
*** [.pio\build\STM32H7-Generic-STM32H750\SrcWrapper\src\HAL\stm32yyxx_hal_adc.c.o] Error 1
<command-line>: fatal error: variant_WeActMiniH7xx.h: No such file or directory

*******************************************************************************
* Looking for variant_WeActMiniH7xx.h dependency? Check our library registry!
*
* CLI  > platformio lib search "header:variant_WeActMiniH7xx.h"
* Web  > https://registry.platformio.org/search?q=header:variant_WeActMiniH7xx.h
*
*******************************************************************************

compilation terminated.
*** [.pio\build\STM32H7-Generic-STM32H750\SrcWrapper\src\HAL\stm32yyxx_hal_adc_ex.c.o] Error 1
<command-line>: fatal error: variant_WeActMiniH7xx.h: No such file or directory

*******************************************************************************
* Looking for variant_WeActMiniH7xx.h dependency? Check our library registry!
*
* CLI  > platformio lib search "header:variant_WeActMiniH7xx.h"
* Web  > https://registry.platformio.org/search?q=header:variant_WeActMiniH7xx.h
*
*******************************************************************************

compilation terminated.
*** [.pio\build\STM32H7-Generic-STM32H750\SrcWrapper\src\HAL\stm32yyxx_hal_cec.c.o] Error 1
<command-line>: fatal error: variant_WeActMiniH7xx.h: No such file or directory

*******************************************************************************
* Looking for variant_WeActMiniH7xx.h dependency? Check our library registry!
*
* CLI  > platformio lib search "header:variant_WeActMiniH7xx.h"
* Web  > https://registry.platformio.org/search?q=header:variant_WeActMiniH7xx.h
*
*******************************************************************************

compilation terminated.
*** [.pio\build\STM32H7-Generic-STM32H750\SrcWrapper\src\HAL\stm32yyxx_hal_comp.c.o] Error 1
========================================================================================= [FAILED] Took 2.97 seconds =========================================================================================

Environment                Status    Duration
-------------------------  --------  ------------
STM32H7-Generic-STM32H750  FAILED    00:00:02.974
==================================================================================== 1 failed, 0 succeeded in 00:00:02.974 ====================================================================================
 *  终端进程“C:\Users\Administrator\.platformio\penv\Scripts\platformio.exe 'run', '--environment', 'STM32H7-Generic-STM32H750'”已终止，退出代码: 1。 
 *  终端将被任务重用，按任意键关闭。 

 *  正在执行任务: C:\Users\Administrator\.platformio\penv\Scripts\platformio.exe run --environment STM32H7-Generic-STM32H750 

Processing STM32H7-Generic-STM32H750 (board: weact_mini_h750vbtx; platform: platformio/ststm32@^17.5.0; framework: arduino)
---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------Verbose mode can be enabled via `-v, --verbose` option
CONFIGURATION: https://docs.platformio.org/page/boards/ststm32/weact_mini_h750vbtx.html
PLATFORM: ST STM32 (17.6.0) > WeAct Studio MiniSTM32H750VBTX
HARDWARE: STM32H750VBT6 480MHz, 128KB RAM, 128KB Flash
DEBUG: Current (cmsis-dap) External (blackmagic, cmsis-dap, jlink, stlink)
PACKAGES:
 - framework-arduinoststm32 @ 4.20801.240815 (2.8.1)
 - framework-cmsis @ 2.50900.0 (5.9.0)
 - tool-dfuutil @ 1.11.0
 - tool-openocd @ 3.1200.0 (12.0)
 - toolchain-gccarmnoneeabi @ 1.120301.0 (12.3.1)
Converting uCNC.ino
Warning! Cannot find linker script for the current target!

LDF: Library Dependency Finder -> https://bit.ly/configure-pio-ldf
LDF Modes: Finder ~ chain, Compatibility ~ soft
Found 14 compatible libraries
Scanning dependencies...
Dependency Graph
|-- TinyUSB @ 0.17.0+sha.5ea637b
|-- SPI @ 1.1.0
|-- Wire @ 1.0.0
Building in release mode
Compiling .pio\build\STM32H7-Generic-STM32H750\SrcWrapper\src\HAL\stm32yyxx_hal.c.o
Compiling .pio\build\STM32H7-Generic-STM32H750\SrcWrapper\src\HAL\stm32yyxx_hal_adc.c.o
Compiling .pio\build\STM32H7-Generic-STM32H750\SrcWrapper\src\HAL\stm32yyxx_hal_adc_ex.c.o
<command-line>: fatal error: variant_WeActMiniH7xx.h: No such file or directory

*******************************************************************************
* Looking for variant_WeActMiniH7xx.h dependency? Check our library registry!
*
* CLI  > platformio lib search "header:variant_WeActMiniH7xx.h"
* Web  > https://registry.platformio.org/search?q=header:variant_WeActMiniH7xx.h
*
*******************************************************************************

compilation terminated.
Compiling .pio\build\STM32H7-Generic-STM32H750\SrcWrapper\src\HAL\stm32yyxx_hal_cec.c.o
*** [.pio\build\STM32H7-Generic-STM32H750\SrcWrapper\src\HAL\stm32yyxx_hal.c.o] Error 1
<command-line>: fatal error: variant_WeActMiniH7xx.h: No such file or directory

*******************************************************************************
* Looking for variant_WeActMiniH7xx.h dependency? Check our library registry!
*
* CLI  > platformio lib search "header:variant_WeActMiniH7xx.h"
* Web  > https://registry.platformio.org/search?q=header:variant_WeActMiniH7xx.h
*
*******************************************************************************

compilation terminated.
*** [.pio\build\STM32H7-Generic-STM32H750\SrcWrapper\src\HAL\stm32yyxx_hal_adc.c.o] Error 1
<command-line>: fatal error: variant_WeActMiniH7xx.h: No such file or directory

*******************************************************************************
* Looking for variant_WeActMiniH7xx.h dependency? Check our library registry!
*
* CLI  > platformio lib search "header:variant_WeActMiniH7xx.h"
* Web  > https://registry.platformio.org/search?q=header:variant_WeActMiniH7xx.h
*
*******************************************************************************

compilation terminated.
*** [.pio\build\STM32H7-Generic-STM32H750\SrcWrapper\src\HAL\stm32yyxx_hal_adc_ex.c.o] Error 1
<command-line>: fatal error: variant_WeActMiniH7xx.h: No such file or directory

*******************************************************************************
* Looking for variant_WeActMiniH7xx.h dependency? Check our library registry!
*
* CLI  > platformio lib search "header:variant_WeActMiniH7xx.h"
* Web  > https://registry.platformio.org/search?q=header:variant_WeActMiniH7xx.h
*
*******************************************************************************

compilation terminated.
*** [.pio\build\STM32H7-Generic-STM32H750\SrcWrapper\src\HAL\stm32yyxx_hal_cec.c.o] Error 1
========================================================================================= [FAILED] Took 2.85 seconds =========================================================================================

Environment                Status    Duration
-------------------------  --------  ------------
STM32H7-Generic-STM32H750  FAILED    00:00:02.846
==================================================================================== 1 failed, 0 succeeded in 00:00:02.846 ====================================================================================
 *  终端进程“C:\Users\Administrator\.platformio\penv\Scripts\platformio.exe 'run', '--environment', 'STM32H7-Generic-STM32H750'”已终止，退出代码: 1。 
 *  终端将被任务重用，按任意键关闭。

**Reply by Paciente8159** [reply]
It seems like your PIO was not did not find or load the board platform resources correctly.
Try deleting the .pio folder on the uCNC directory and try again.
If that does not work try deleting all your platformio packages and forcing them to be re-downloaded.
